internal/provider/setup_key_resource.go

```diff
@@ -0,0 +1,348 @@
+		Ephemeral:           setupKey.Ephemeral,
+		Expires:             setupKey.Expires,
+		Id:                  setupKey.Id,
+		Key:                 setupKey.Key,
+		LastUsed:            setupKey.LastUsed,
+		Name:                setupKey.Name,
+		Revoked:             setupKey.Revoked,
+		State:               setupKey.State,
+		Type:                setupKey.Type,
+		UpdatedAt:           setupKey.UpdatedAt,
+		UsageLimit:          setupKey.UsageLimit,
+		UsedTimes:           setupKey.UsedTimes,
+		Valid:               setupKey.Valid,
+	}, &data)...)
+	data.Key = types.StringValue(setupKey.Key)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Save data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *SetupKey) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data SetupKeyModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	setupKey, err := r.client.SetupKeys.Get(ctx, data.Id.ValueString())
+
+	if err != nil {
+		if strings.Contains(err.Error(), "not found") {
+			data.Id = types.StringNull()
+			resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+		} else {
+			resp.Diagnostics.AddError("Error getting SetupKey", err.Error())
+		}
+		return
+	}
+
+	resp.Diagnostics.Append(setupKeyAPIToTerraform(ctx, setupKey, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *SetupKey) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var data SetupKeyModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	if data.Id.ValueString() == "" {
+		r.Create(ctx, resource.CreateRequest{Config: req.Config, Plan: req.Plan, ProviderMeta: req.Config}, (*resource.CreateResponse)(resp))
+		return
+	}
+
+	setupKey, err := r.client.SetupKeys.Update(ctx, data.Id.ValueString(), api.SetupKeyRequest{
+		AutoGroups: stringListDefault(ctx, data.AutoGroups, []string{}),
+		Revoked:    data.Revoked.ValueBool(),
+	})
+
+	if err != nil {
+		resp.Diagnostics.AddError("Error updating User", err.Error())
+		return
+	}
+
+	resp.Diagnostics.Append(setupKeyAPIToTerraform(ctx, setupKey, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+
+}
+
+func (r *SetupKey) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var data SetupKeyModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	err := r.client.SetupKeys.Delete(ctx, data.Id.ValueString())
```

Comment:
Should probably support `revoke` in addition to `delete` resource deletion flow.

Reply:
How would that work given terraform only supports create/update/delete? If a user wants to revoke a setup key they should update the `revoked` field instead?